File: 0015-3sum/0015-3sum.cpp
class Solution {
  void twoSumII(vector<int>& nums, int i, vector<vector<int>>& res) {
    int lo = i + 1, hi = nums.size() - 1;
    while (lo < hi) {
      int sum = nums[i] + nums[lo] + nums[hi];
      if (sum < 0) {
        ++lo;
      } else if (sum > 0) {
        --hi;
      } else {
        res.push_back({nums[i], nums[lo++], nums[hi--]});
        while (lo < hi && nums[lo] == nums[lo - 1]) ++lo;
      }
    }
  }

 public:
  vector<vector<int>> threeSum(vector<int>& nums) {
    vector<vector<int>> output;
    sort(nums.begin(), nums.end());
    for (int i = 0; i < nums.size() && nums[i] <= 0; i++) {
      if (i == 0 || nums[i - 1] != nums[i]) {
          twoSumII(nums, i, output);
      }
    }
    return output;
  }
};

File: 0015-3sum/0015-3sum.py
# time complexity: O(n^2)
# space complexity: O(n)
from typing import List


class Solution:
    def twoSum(self, nums: List[int], i: int, result: List[List[int]]):
        low = i + 1
        hight = len(nums) - 1
        while (low < hight):
            sum = nums[i] + nums[low] + nums[hight]
            if sum < 0:
                low += 1
            elif sum > 0:
                hight -= 1
            else:
                result.append([nums[i], nums[low], nums[hight]])
                low += 1
                hight -= 1
                while low < hight and nums[low] == nums[low - 1]:
                    low += 1

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()
        for i, item in enumerate(nums):
            if item > 0:
                break
            if item != nums[i-1] or i == 0:
                self.twoSum(nums, i, result)

        return result
